src/Java.Interop/Java.Interop/JniRuntime.cs

```diff
@@ -403,21 +402,32 @@ public void DestroyRuntime ()
 
 		internal void Track (JniType value)
 		{
-			TrackedInstances.TryAdd (value.PeerReference.Handle, value);
+			lock (TrackedInstances) {
+				if (!TrackedInstances.ContainsKey (value.PeerReference.Handle))
+					TrackedInstances [value.PeerReference.Handle] = value;
+			}
 		}
 
 		internal void UnTrack (IntPtr key)
 		{
-			TrackedInstances.TryRemove (key, out var _);
+			lock (TrackedInstances) {
+				if (TrackedInstances.ContainsKey (key))
+					TrackedInstances.Remove (key);
+			}
 		}
 
 		void ClearTrackedReferences ()
 		{
-			foreach (var k in TrackedInstances.Keys.ToList ()) {
-				if (TrackedInstances.TryRemove (k, out var d))
-					d.Dispose ();
+			lock (TrackedInstances) {
+				foreach (var k in TrackedInstances.Keys.ToList ()) {
+					if (TrackedInstances.TryGetValue (k, out var d)) {
+						TrackedInstances.Remove (k);
+						d.Dispose ();
```

Comment:
"General" threading advice I've run across is "don't invoke 'unknown code' from within a lock".  This `d.Dispose()` code is "unknown code", and thus shouldn't be executed within the lock.

Thus, we perhaps should prefer:

```csharp
var disposees = new List<IDisposable>();
lock (TrackedInstances) {
    foreach (var e in TrackedInstances) {
        disposees.Add (e.Value);
    }
    TrackedInstances.Clear ();
}
foreach (var d in disposees) {
    try {
        d.Dispose ();
    }
    catch (Exception e) {
        // log…?
    }
}
```

Reply:
Good catch. I updated the method. Wasn't sure about reasons to catch the exceptions here (what to log and whether to rethrow or throw different exception?), so I left it without catch block, same as before the change.